lib/framework/procbot.ts

```diff
@@ -36,8 +46,8 @@ const exec: (command: string, options?: any) => Promise<{}> = Promise.promisify(
 export class ProcBot {
 	protected _botname: string;
 	protected logger = new Logger();
-	private emitters: ServiceEmitter[] = [];
-	private listeners: ServiceListener[] = [];
+	private listeners: Map<string, ServiceMap> = new Map<string, ServiceMap>();
+	private emitters: Map<string, ServiceMap> = new Map<string, ServiceMap>();
```

Comment:
Could we make `ServiceMap` generic in its `.instance` type, so that these maps are known to only contain emitters or listeners up front, and we don't need to cast the results later?